### ISSUE TITLE
IDP4 and MVC 5.0, ProtocolValidator.RequireNonce error on Google Chrome

### ISSUE THREAD

**Issue by felipeoriani** [skoruba/IdentityServer4.Admin#555]
Hey guys. We have an application made with ASP.NET MVC 5 that will be part of a solution to authenticate using a Single Sign On made with Identity Server 4 and the Skoruba implementation with Asp.Net Core 2.2. It is working on Microsoft Edge and Firefox but on the Google Chrome we are getting the error message with

> IDX21323: RequireNonce is '[PII is hidden]'. OpenIdConnectProtocolValidationContext.Nonce was null, OpenIdConnectProtocol.ValidatedIdToken.Payload.Nonce was not null. The nonce cannot be validated. If you don't need to check the nonce, set OpenIdConnectProtocolValidator.RequireNonce to 'false'. Note if a 'nonce' is found it will be evaluated.

I as searching on web about the error and I have seen that Chrome is not able to read an external cookie for authentication but I am not sure if I understand correctly. I have tried to add the `app.UseExternalSignInCookie(..)` method to make sure that the application could be able to read it but no good results so far. The code bellow, is what we have made to implement the SSO on ASP.NET MVC 5 authentication:

```

private static void UseSingleSignOnAuthentication(this IAppBuilder app)
{
    JwtSecurityTokenHandler.DefaultInboundClaimTypeMap.Clear();

    app.UseCookieAuthentication(new CookieAuthenticationOptions
    {
        AuthenticationType = DefaultAuthenticationTypes.ApplicationCookie
    });

    app.UseExternalSignInCookie(DefaultAuthenticationTypes.ExternalCookie);

    app.UseOpenIdConnectAuthentication(new OpenIdConnectAuthenticationOptions
    {
        ClientId = "..", // the client id generated on Skoruba portal
        ClientSecret = "..", // the client secret generated on Skoruba portal
        Authority = "..", // the base address to the STS Skoruba
        RedirectUri = "..", // Url of this application
        PostLogoutRedirectUri = "..", // Url of this application
        ResponseType = "code id_token",
        Scope = "openid email profile",
        RequireHttpsMetadata = false,
        UseTokenLifetime = false,
        SignInAsAuthenticationType = "Cookies",
        SaveTokens = true,

        Notifications = new OpenIdConnectAuthenticationNotifications
        {
            AuthorizationCodeReceived = async notification =>
            {
                var tokenClient = new TokenClient(SsoSettings.TokenEndpoint,
                    SsoSettings.ClientId,
                    SsoSettings.ClientSecret);

                var tokenResponse = await tokenClient.RequestAuthorizationCodeAsync(notification.Code, notification.RedirectUri);

                var userInfoClient = new UserInfoClient(SsoSettings.UserInfoEndpoint);
                var userInfoResponse = await userInfoClient.GetAsync(tokenResponse.AccessToken);

                var auth = userInfoResponse.Auth(tokenResponse, notification);
                AuthHelper.FillUpUserInfo(auth);

                // authenticate user with Owin and define claims from user and SSO based on userInfoResponse
                var user = AuthHelper.Login(tokenResponse, userInfoResponse);

                if (user == null)
                    thrown newException ( .... );
            },
            RedirectToIdentityProvider = n => { /* logout code */ }
        }
    });
}
```
I have tried it on serveral machines with Chrome to make sure that is not something on a single machine and I get the same error. Does anyone already pass to it on Google Chrome and know how to fix it? 

Thank you.

**Comment by skoruba** [comment]
Hi @felipeoriani - sorry for delay, how did you fix it? Everything work?
Thx

**Comment by felipeoriani** [comment]
Hi @skoruba, yes, we have added https on the client (asp.net mvc 5) and now it works, but we still have a problem in Chrome, we can't logout, even on the admin panel. If we click at logout button, the application just login again. The same happen for client apps. On the other browsers it works (Edge and Firefox).

Is there any way to fix it on the STS application?

**Comment by khaninejad** [comment]
has same issue

**Comment by skoruba** [comment]
Please, could you please switch to dev branch and try install project with the newest RC2?
https://github.com/skoruba/IdentityServer4.Admin/tree/dev

**Comment by khaninejad** [comment]
@skoruba fixed by running this command on windows run:

`Chrome --disable-web-security --user-data-dir=~/ChromeUserData/`

**Comment by felipeoriani** [comment]
@khaninejad and @skoruba  I have been facing this intermittent problem of IDX21323 with some users and given it is intermittent we are not able to investigate it properly. Is there any User Friendly way to change this configuration the @khaninejad suggested here in Google Chrome or the machine? I want to avoid to ask the users running this command on the prompt as administrator.

@skoruba was it fixed in the lastest versions of your project?

Thank you.

**Comment by khaninejad** [comment]
apparently this bug related to Content-Security-Policy.

**Comment by skoruba** [comment]
@khaninejad - do you have more details about this? Thanks.

**Comment by felipeoriani** [comment]
Hey guys,

Considering it is intermittent problem with some users, I have found a way to simulate it. Try to following these steps:

- Go to Chrome
- On url enter with `chrome://flags`
- On the search text look for `Cookies`
- Set the flag `Cookies without SameSite must be secure` to `Disabled`
- Restart chrome

Then if you try to access a client from the project, you will get the error of `IDX21323` of the Required Nonce. As we know, the browser new Edge uses the same engine of the Chrome and this problem does not happen over there. Looking for why it is happening with Chrome, I have found the following blog post: 
https://www.thinktecture.com/en/identity/samesite/prepare-your-identityserver/

We are trying to get a way to make sure our clients works and maybe it can help us to fix the problem (on @skoruba project and client projects).
